{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "fd083f42",
   "metadata": {},
   "outputs": [],
   "source": [
    "# it converts the safety windows intervals into safety windows using the reference as source\n",
    "def safetyWindows(data):\n",
    "    \n",
    "    nS = data['sequences per cluster']\n",
    "    R = data['reference']\n",
    "    intervals = data['safety windows intervals']\n",
    "    \n",
    "    safetyWindows = {}\n",
    "    refSeq = [char for char in R]\n",
    "    refPos = [0]*len(refSeq)\n",
    "    \n",
    "    for s in range(1,nS):\n",
    "        intSeq = intervals[s]\n",
    "        windowsList = list()\n",
    "        for n in range(0,len(intSeq)):\n",
    "            intB = int(intSeq[n][0])\n",
    "            intE = int(intSeq[n][1])-1\n",
    "            windowsList.append(''.join(refSeq[intB:(intE)]))\n",
    "        safetyWindows[s] = windowsList\n",
    "    \n",
    "    data['safety windows'] = safetyWindows\n",
    "    \n",
    "    return data;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6b6dc960",
   "metadata": {},
   "outputs": [],
   "source": [
    "def overlappingWindowsSize(data):\n",
    "    \n",
    "    # removing overlapping details and oversimplify the windows\n",
    "    \n",
    "    safetyIntervals = data['safety windows intervals'] # safety windows intervals\n",
    "    nS = data['number of sequences'] # number of stirngs\n",
    "    overlaps = [0]*nS \n",
    "    overlappingLength = [0]*nS\n",
    "    \n",
    "    for i in range(1,nS):\n",
    "        arrayIntervals = [0]*(len(safetyIntervals[i])*2)\n",
    "        currentInterval = safetyIntervals[i]\n",
    "        cnt = 0\n",
    "        \n",
    "        for k in range(0,len(currentInterval)):\n",
    "            arrayIntervals[cnt] = int(currentInterval[k][0])\n",
    "            arrayIntervals[cnt+1] = int(currentInterval[k][1])\n",
    "            cnt += 2\n",
    "        \n",
    "        # calculate the number of overlapping cases;\n",
    "        if (all(x <= y for x,y in zip(arrayIntervals, arrayIntervals[1:])) == False):\n",
    "            overlaps[i] = 1\n",
    "        \n",
    "        # remove the overlapping cases in the sequences and using that information to correct length and coverage\n",
    "        if overlaps[i] != 0:\n",
    "            arrayPairIntervals = list(split_pairs(arrayIntervals[0:len(arrayIntervals)+1], 2))\n",
    "\n",
    "            mappingPositions = [-1]*(max(arrayIntervals))\n",
    "\n",
    "            for j in arrayPairIntervals:\n",
    "                for l in range(j[0],j[1]):\n",
    "                    mappingPositions[l] += 1 \n",
    "\n",
    "            overlappingLength[i] = sum(mappingPositions[l] for l in range(0,max(arrayIntervals)))\n",
    "                                       \n",
    "        \n",
    "    \n",
    "    data['overlap'] = overlaps\n",
    "    data['overlapping length'] = overlappingLength\n",
    "    \n",
    "    \n",
    "    return data;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "67943315",
   "metadata": {},
   "outputs": [],
   "source": [
    "# calculate the length of the safety windows in that subcluster\n",
    "# calculate the coverage of the safety windows in that subcluster\n",
    "def coverageAverage(data):\n",
    "    \n",
    "    tuples = data['tuples']\n",
    "    safetyWindows = data['safety windows']\n",
    "    R = data['reference']\n",
    "    seqComp = data['sequences']\n",
    "    overlappingLength = data['overlapping length']\n",
    "    \n",
    "    lengthPerSubCluster = [0]*len(tuples)\n",
    "    coveragePerSubCluster= [0]*len(tuples)\n",
    "    \n",
    "    # calculate such values for all tuples based on the criteria that were chosen;\n",
    "    for i in range(0,len(tuples)):\n",
    "        lengthPerTuple = 0\n",
    "        coveragePerTuple = 0\n",
    "        for j in range(0,len(tuples[i])):\n",
    "            windowSeq = 0\n",
    "            elementTuple = tuples[i][j]\n",
    "            seqID = elementTuple\n",
    "            windowsPerSeqID = safetyWindows[seqID]\n",
    "            windowsLength = -(overlappingLength[elementTuple])\n",
    "\n",
    "            for k in range(0,len(windowsPerSeqID)):\n",
    "                windowsLength += len(windowsPerSeqID[k])\n",
    "\n",
    "            coveragePerTuple += (windowsLength/len(R)) \n",
    "            lengthPerTuple += windowsLength\n",
    "        \n",
    "        \n",
    "        coveragePerSubCluster[i] = coveragePerTuple/3  \n",
    "        lengthPerSubCluster[i] = lengthPerTuple/3\n",
    "        \n",
    "        \n",
    "    # save new values into the data structure\n",
    "    data['average length'] = lengthPerSubCluster \n",
    "    data['coverage average']= coveragePerSubCluster\n",
    "        \n",
    "    return data;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "cd377495",
   "metadata": {},
   "outputs": [],
   "source": [
    "# process tuples calculation\n",
    "import networkx as nx\n",
    "import itertools\n",
    "KT = 3\n",
    "from ipynb.fs.full.commonFunctions import *\n",
    "def TupleAverage(data):\n",
    "    \n",
    "    tuples = data['tuples'] # read tuples create in the file\n",
    "    safetyWindows = data['safety windows'] # read safety windows\n",
    "    \n",
    "    graphLength = [0]*len(tuples)\n",
    "    graphNoise = [0]*len(tuples)\n",
    "    numberCliques = [0]*len(tuples)\n",
    "    graphCoverage = [0]*len(tuples)\n",
    "    \n",
    "    referenceLength = len(data['reference'])\n",
    "    cliquesPerTuples = {}\n",
    "\n",
    "    # For each tuple, it creates a subgraph\n",
    "    for i in range(0,len(tuples)):\n",
    "        currentTuple = tuples[i]\n",
    "        currentGraph = nx.Graph()\n",
    "\n",
    "        \n",
    "        for j in range(0,len(currentTuple)):\n",
    "            currentElement = currentTuple[j]\n",
    "\n",
    "            windowsPerSeqID = safetyWindows[currentElement]\n",
    "            for k in range(0,len(windowsPerSeqID)):\n",
    "                if windowsPerSeqID[k] != \"\":\n",
    "                    currentGraph.add_node(windowsPerSeqID[k]+\" \"+str(currentElement))\n",
    "            \n",
    "            \n",
    "            # same process is done for tuples as it was done for cluster\n",
    "            pairCluster = list(itertools.combinations(tuples[i],2))\n",
    "            for (c1,c2) in pairCluster:\n",
    "\n",
    "                windowsInC1 = safetyWindows[c1]\n",
    "                windowsInC2 = safetyWindows[c2]\n",
    "                \n",
    "                for k in range(0,len(windowsInC1)):\n",
    "                    for l in range(0,len(windowsInC2)):\n",
    "                        matchStr = longestMatchingSubstring(windowsInC1[k],windowsInC2[l])\n",
    "                        maxLength = max(len(windowsInC1[k]),len(windowsInC2[l]))\n",
    "                        if len(matchStr) >= 0.5*maxLength:\n",
    "                            if windowsInC1[k] != \"\" or windowsInC2[l] !=\"\":\n",
    "                                currentGraph.add_edge(windowsInC1[k]+\" \"+str(c1),windowsInC2[l]+\" \"+str(c2))\n",
    "            \n",
    "        CliquesInCluster = list(nx.find_cliques(currentGraph))\n",
    "        cliquesPerTuples[i] = CliquesInCluster\n",
    "        \n",
    "        currentLength = 0\n",
    "        cliqueCounter = 0\n",
    "        noise = 0 \n",
    "        for clique in CliquesInCluster:\n",
    "            if len(clique) >= KT:\n",
    "                cliqueCounter += 1\n",
    "                readInClique = ['']*len(clique)\n",
    "                for k in range(0,len(clique)):\n",
    "                    readInClique[k] = clique[k].split(\" \")[0]\n",
    "                \n",
    "                maxReadInClique = maxMatchingString(readInClique)\n",
    "                \n",
    "                for k in range(0,len(clique)):\n",
    "                    noise += abs(len(readInClique[k]) - len(maxReadInClique))/referenceLength\n",
    "                    \n",
    "                currentLength += len(maxReadInClique)\n",
    "        \n",
    "        if cliqueCounter == 0:\n",
    "            graphLength[i] = 0\n",
    "            numberCliques[i] = 0\n",
    "            graphNoise[i] = 0\n",
    "        else:\n",
    "            graphCoverage[i] = (currentLength/cliqueCounter)/referenceLength\n",
    "            graphLength[i] = currentLength\n",
    "            numberCliques[i] = cliqueCounter\n",
    "            graphNoise[i] = noise/cliqueCounter\n",
    "    \n",
    "    \n",
    "    data['cliques per tuples'] = cliquesPerTuples\n",
    "    data['length of cliques per tuples'] = graphLength # save the length of safety windows common in the tuple\n",
    "    data['number of cliques per tuples'] = numberCliques # save the coverage of safety windows common in the tuple\n",
    "    data['noise of cliques per tuples'] = graphNoise\n",
    "    data['coverage of cliques per tuples'] = graphCoverage\n",
    "    \n",
    "    return data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "5011402b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create a safety graphs\n",
    "def safetyGraph(data):\n",
    "    \n",
    "    nS = data['sequences per cluster']\n",
    "    R = data['reference']\n",
    "    intervals = data['safety windows intervals']\n",
    "    safetyWindows = data['safety windows']\n",
    "    \n",
    "    # building graph\n",
    "    \n",
    "    cliques = {}\n",
    "    \n",
    "        \n",
    "    Graph = nx.Graph()\n",
    "\n",
    "    for s in range(1,nS):\n",
    "        windowRef = safetyWindows[s]\n",
    "        for k in range(0,len(windowRef)):\n",
    "            Graph.add_node(windowRef[k]) # add nodes based on the safety windows in the sequence\n",
    "\n",
    "    seqClu= [i for i in range(1,nS)]\n",
    "    pairClu = list(itertools.combinations(seqClu,2))\n",
    "    for (c1,c2) in pairClu:\n",
    "        sizec1 = len(intervals[c1])\n",
    "        sizec2 = len(intervals[c2])\n",
    "        C1 = safetyWindows[c1]\n",
    "        C2 = safetyWindows[c2]\n",
    "        for j in range(0,sizec1):\n",
    "            for k in range(0,sizec2):\n",
    "                matchStr = longestMatchingSubstring(C1[j],C2[k])\n",
    "                maxLength = max(len(C1[j]),len(C2[k]))\n",
    "                if len(matchStr) > 0.5*maxLength:\n",
    "                    Graph.add_edge(C1[j],C2[k],weight = 1)\n",
    "                    \n",
    "\n",
    "\n",
    "    cliques = list(nx.find_cliques(Graph))\n",
    "    \n",
    "    ratioLengthPerClique = [0]*len(cliques)\n",
    "    ratioCoveragePerClique = [0]*len(cliques)\n",
    "    ratioNoisePerClique = [0]*len(cliques)\n",
    "    for k in range(0,len(cliques)):\n",
    "        if len(cliques[k]) >= 0:\n",
    "            emptyString = 0\n",
    "            for a in cliques[k]:\n",
    "                if a==\"\":\n",
    "                    emptyString = 1\n",
    "            if emptyString ==0:\n",
    "                maxString = maxMatchingString(cliques[k])\n",
    "                ratioCoveragePerClique[k] = sum([len(maxString)/len(element) for element in cliques[k]])/len(cliques[k])\n",
    "                ratioLengthPerClique[k] = sum([len(maxString) for element in cliques[k]])/len(cliques[k])\n",
    "                ratioNoisePerClique[k] = 1 - ratioCoveragePerClique[k]\n",
    "    \n",
    "    data['interval cliques'] = cliques\n",
    "    data['ratio length'] = ratioLengthPerClique\n",
    "    data['ratio coverage'] = ratioCoveragePerClique\n",
    "    data['ratio noise'] = ratioNoisePerClique\n",
    "    \n",
    "    \n",
    "    return data;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "955be37e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def subCliques(data):\n",
    "    \n",
    "    tupleCliques = data['cliques per tuples']\n",
    "    referenceLength = len(data['reference'])\n",
    "    \n",
    "    cliqueNonMaxPerTuple = [0]*len(tupleCliques)\n",
    "    nonMaxAverageLengthPerTuple = [0]*len(tupleCliques)\n",
    "    nonMaxAverageCoveragePerTuple = [0]*len(tupleCliques)\n",
    "    nonMaxAverageNoisePerTuple = [0]*len(tupleCliques)\n",
    "    \n",
    "    nonMaxAverageLength = 0\n",
    "    nonMaxAverageCoverage = 0\n",
    "    nonMaxAverageNoise = 0\n",
    "    \n",
    "    # calculation for cliques and overall\n",
    "    \n",
    "    for i in range(0,len(tupleCliques)):\n",
    "        for j in range(0,len(tupleCliques[i])):\n",
    "            if len(tupleCliques[i][j]) < KT:\n",
    "                cliqueNonMaxPerTuple[i] += 1\n",
    "        \n",
    "        windowsInClique = list()\n",
    "        for j in range(0,len(tupleCliques[i])):\n",
    "            if len(tupleCliques[i][j]) < KT:\n",
    "                for a in tupleCliques[i][j]:\n",
    "                    windowsInClique.append(a.split(' ')[0])\n",
    "            \n",
    "        if len(windowsInClique) < KT:\n",
    "            lengthInSafetyWindows = 0\n",
    "            coverageInSafetyWindows = 0\n",
    "            noiseInSafetyWindows = 0\n",
    "            maxRead = maxMatchingString(windowsInClique)\n",
    "            \n",
    "            for windows in windowsInClique:\n",
    "                lengthInSafetyWindows += len(windows)/(len(windowsInClique))\n",
    "                nonMaxAverageLength += len(windows)\n",
    "                coverageInSafetyWindows += len(windows)/referenceLength/(len(windowsInClique))\n",
    "                nonMaxAverageCoverage += len(windows)/referenceLength\n",
    "                noiseInSafetyWindows += (1 - coverageInSafetyWindows)/(len(windowsInClique))\n",
    "                nonMaxAverageNoise += (1 - coverageInSafetyWindows)\n",
    "        \n",
    "        if cliqueNonMaxPerTuple[i] != 0:\n",
    "            #print(cliqueNonMax[i])\n",
    "            nonMaxAverageLengthPerTuple[i] = (lengthInSafetyWindows/cliqueNonMaxPerTuple[i])\n",
    "            nonMaxAverageCoveragePerTuple[i] = (coverageInSafetyWindows/cliqueNonMaxPerTuple[i])\n",
    "            nonMaxAverageNoisePerTuple[i] = (noiseInSafetyWindows/cliqueNonMaxPerTuple[i])\n",
    "        \n",
    "    \n",
    "    # calculation of amount of small safety windows\n",
    "    cliqueNonMaxTotal = sum(cliqueNonMaxPerTuple[i] for i in range(0,len(cliqueNonMaxPerTuple)))\n",
    "    nonMaxAverageLength = nonMaxAverageLength/cliqueNonMaxTotal\n",
    "    nonMaxAverageCoverage = nonMaxAverageCoverage/cliqueNonMaxTotal\n",
    "    nonMaxAverageNoise = nonMaxAverageNoise/cliqueNonMaxTotal\n",
    "    \n",
    "    \n",
    "    data['cliqueNonMaxPerTuple'] = cliqueNonMaxPerTuple\n",
    "    data['cliqueNonMaxTotal'] = cliqueNonMaxTotal\n",
    "    data['nonMaxAverageLengthPerTuple'] = nonMaxAverageLengthPerTuple\n",
    "    data['nonMaxAverageCoveragePerTuple'] = nonMaxAverageCoveragePerTuple\n",
    "    data['nonMaxAverageNoisePerTuple'] = nonMaxAverageNoisePerTuple\n",
    "    data['nonMaxAverageLength'] = nonMaxAverageLength\n",
    "    data['nonMaxAverageCoverage'] = nonMaxAverageCoverage\n",
    "    data['nonMaxAverageNoise'] = nonMaxAverageNoise\n",
    "        \n",
    "    return data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d11b797a",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "def createTuples(data,path,subcluster,k):\n",
    "    \n",
    "    # getting the correct order\n",
    "    file = path+'/fasta/'+subcluster+'.fasta'\n",
    "    cnt = 0\n",
    "    \n",
    "    # retrieve the sequence in the fasta file\n",
    "    sequenceInFast = {}\n",
    "    proteinSequence = set()\n",
    "    taxonomyID = list()\n",
    "    indicatorID = list()\n",
    "    \n",
    "    with open(file) as fp:\n",
    "        for line in fp:\n",
    "            if line == '\\n':\n",
    "                break\n",
    "            linevector = line.split()\n",
    "            if (\">\" in linevector[0])==True :\n",
    "                sequenceInFast[cnt] = linevector[0].split('|')[1]\n",
    "                cnt += 1\n",
    "                \n",
    "    fasta_sequences = SeqIO.parse(open(file),'fasta')\n",
    "    for fasta in fasta_sequences:\n",
    "        name, sequence,trials = fasta.id, str(fasta.seq),[piece for piece in fasta.description.split(\"|\")[2].split(\" \") if \"OX=\" in piece]\n",
    "        taxonomyID.append(trials[0].split(\"=\")[1])\n",
    "        indicatorID.append(name.split(\"|\")[1])\n",
    "        proteinSequence.add(sequence)\n",
    "        \n",
    "    sequences = data['sequences ID']\n",
    "    \n",
    "    # choose criteria to experiment with\n",
    "    escore = data['e-value full']\n",
    "    escoreByFasta = {}\n",
    "    for i in range(1,len(sequenceInFast)):\n",
    "        index = list(sequences.keys())[list(sequences.values()).index(sequenceInFast[i])]\n",
    "        escoreByFasta[i] = escore[index]\n",
    "    \n",
    "    escoreByFasta = dict(sorted(escoreByFasta.items(), key=lambda item: item[1]))\n",
    "    \n",
    "    # order the sequences in the cluster based on the criteria chosen\n",
    "    arr = list(escoreByFasta.keys())\n",
    "    results = [arr[i:i+k] for i in range(len(arr)-k+1)]\n",
    "    \n",
    "    data['tuples'] = results\n",
    "    data['protein sequence'] = proteinSequence\n",
    "    data['taxonomy ID'] = taxonomyID\n",
    "    data['indicator ID'] = indicatorID\n",
    "    \n",
    "    return data"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
